src/main/surfrdUtilsMod.F90

```diff
@@ -143,6 +145,205 @@ subroutine convert_cft_to_pft( begg, endg, cftsize, wt_cft )
 
   end subroutine convert_cft_to_pft
 
+  !-----------------------------------------------------------------------
+  subroutine collapse_all_pfts(wt_lunit, wt_nat_patch, natpft_size, wt_cft, cft_size, begg, endg, n_dom_soil_patches)
+    !
+    ! DESCRIPTION
+    ! Collapse to the top N dominant soil patches (n_dom_soil_patches) among all
+    ! present pfts, cfts, & bare ground.
+    ! - Bare ground could be up to 1 patch before collapsing.
+    ! - Pfts could be up to 14 before collapsing.
+    ! - Cfts could be up to 2 or 78 with use_crop = .F. or .T., respectively.
+    !
+    ! !USES:
+    use clm_varpar, only: natpft_lb, natpft_ub, cft_lb, cft_ub, maxveg
+    use landunit_varcon, only: istsoil, istcrop, max_lunit
+    use pftconMod, only: noveg, nc3crop, pftcon
+    use array_utils, only: find_k_max_indices
+    !
+    ! !ARGUMENTS:
+    ! Use begg and endg rather than 'bounds', because bounds may not be
+    ! available yet when this is called
+    integer, intent(in) :: begg  ! Beginning grid cell index
+    integer, intent(in) :: endg  ! Ending grid cell index
+    integer, intent(in) :: natpft_size  ! CFT size
+    integer, intent(in) :: cft_size  ! CFT size
+    integer, intent(in) :: n_dom_soil_patches  ! # dominant soil patches
+    ! These arrays modified in-place
+    ! Weights of landunits
+    real(r8), intent(inout) :: wt_lunit(begg:,:)
+    ! Weights of managed (cft) and unmanaged (nat_patch) soil patches per
+    ! grid cell
+    ! Dimensioned [g, natpft_lb:natpft_ub]
+    real(r8), intent(inout) :: wt_nat_patch(begg:, natpft_lb:)
+    ! Dimensioned [g, cft_lb:cft_lb+cft_size-1]
+    real(r8), intent(inout) :: wt_cft(begg:, cft_lb:)
+    !
+    ! !LOCAL VARIABLES:
+    integer :: g  ! gridcell index
+    integer :: m  ! soil patch index
+    integer :: n  ! index of the order of the dominant soil patches
+    integer, allocatable :: max_indices(:)  ! array of dominant soil patch index values
+    real(r8) :: wt_all_patch(natpft_lb:cft_ub)
+    real(r8) :: wt_dom_nat_sum
+    real(r8) :: wt_dom_cft_sum
+    character(len=*), parameter :: subname = 'collapse_all_pfts'
+
+    !-----------------------------------------------------------------------
+    SHR_ASSERT_ALL((ubound(wt_lunit) == (/endg, max_lunit/)), errMsg(sourcefile, __LINE__))
+    SHR_ASSERT_ALL((ubound(wt_nat_patch) == (/endg, natpft_lb+natpft_size-1/)), errMsg(sourcefile, __LINE__))
+    SHR_ASSERT_ALL((ubound(wt_cft) == (/endg, cft_lb+cft_size-1/)), errMsg(sourcefile, __LINE__))
+    SHR_ASSERT_ALL((ubound(wt_all_patch) == (/natpft_lb+natpft_size+cft_size-1/)), errMsg(sourcefile, __LINE__))
+
+    ! Find the top N dominant soil patches to collapse pft data to
+    ! n_dom_soil_patches < 0 is not allowed (error check in controlMod.F90)
+    ! n_dom_soil_patches == 0 or 78 currently mean "do not collapse pfts"
+    ! so skip over this subroutine's work
+    if (n_dom_soil_patches > 0 .and. n_dom_soil_patches < 78) then
```

Comment:
Don't use a hardcoded magic number (78). Use an integer variable or parameter set to the specific meaning meant here.

Reply:
Done.